lib/wallaroo/core/source/connector_source/connector_framed_source_notify.pony

```diff
@@ -220,3 +684,267 @@ class ConnectorSourceNotify[In: Any val]
+    Invariant(checkpoint_id == _barrier_checkpoint_id)
+
+    if _session_active then
+      _process_acks_for_active(source)
+
+      // process acks for EOS/_pending_close streams
+      _process_acks_for_pending_close(source)
+      // process any stream relinquish requests
+      _relinquish_streams()
+    end
+
+  /////////////////////////
+  // Local state management
+  /////////////////////////
+  fun ref _process_acks_for_active(source: ConnectorSource[In] ref) =>
+    for (stream_id, s) in _active_streams.pairs() do
+      _pending_acks.push((stream_id, s.last_acked))
+    end
+    if (_fsm_state is _ProtoFsmStreaming) or
+       (_fsm_state is _ProtoFsmShrinking) then
+      _send_acks(source)
+    end
+
+  fun ref _process_acks_for_pending_close(source: ConnectorSource[In] ref) =>
+    for (stream_id, s) in _pending_close.pairs() do
+      ifdef "trace" then
+        @printf[I32]("%s ::: Processing ack for %s at %s\n".cstring(),
+          WallClock.seconds().string().cstring(),
+          stream_id.string().cstring(), s.last_acked.string().cstring())
+      end
+      _pending_acks.push((stream_id, s.last_acked))
+      _pending_relinquish.push(StreamTuple(s.id, s.name, s.last_acked))
+    end
+    // clear _pending_close
+    _pending_close.clear()
+    if (_fsm_state is _ProtoFsmStreaming) or
+       (_fsm_state is _ProtoFsmShrinking) then
+      _send_acks(source)
+    end
+
+  fun ref _relinquish_streams() =>
+    let streams = recover iso Array[StreamTuple] end
+    for s in _pending_relinquish.values() do
+      streams.push(s)
+    end
+    _pending_relinquish.clear()
+
+    if streams.size() > 0 then
+      @printf[I32]("ConnectorSource relinquishing %s streams\n".cstring(),
+        streams.size().string().cstring())
+      _listener.streams_relinquish(source_id, consume streams)
+    else
+      if _fsm_state is _ProtoFsmShrinking then
+        @printf[I32]("ConnectorSource shrinking %s streams\n".cstring(),
+          streams.size().string().cstring())
+        _listener.streams_relinquish(source_id, consume streams)
+      end
+    end
+
+  fun ref _clear_streams() =>
+    _pending_notify.clear()
+    _active_streams.clear()
+    _pending_close.clear()
+    _pending_relinquish.clear()
+
+  fun ref _clear_and_relinquish_all() =>
+    for s_map in [_active_streams ; _pending_close].values() do
+      for s in s_map.values() do
+        _pending_relinquish.push(StreamTuple(s.id, s.name, s.last_seen))
+      end
+    end
+    _relinquish_streams()
+    _clear_streams()
+
+  /////////
+  // Notify
+  /////////
+
+  fun ref _process_notify(source: ConnectorSource[In] ref, stream_id: StreamId,
+    stream_name: String, point_of_reference: PointOfReference)
+  =>
+    // add to _pending_notify
+    _pending_notify.set(stream_id)
+
+    // create a promise for handling result
+    let promise = Promise[NotifyResult[In]]
+    promise.next[None](recover NotifyResultFulfill[In](stream_id, _session_id) end)
```

Comment:
Why can't this be:
```
promise.next[None](_connector_source~stream_notify_result(_session_id))
```
and then `ConnectorSource.stream_notify_result` takes an object that has the success and stream info on it (which could have a more descriptive name than `NotifyResult` and doesn't need a type parameter)?

Reply:
@JONBRWN We can push this off as a refactor with an issue to make sure we get the functionality in sooner.

Reply:
issue opened #2825